### ISSUE TITLE
Image Upload using Drag & Drop

### ISSUE THREAD

**Issue by praveenscience** [jbt/markdown-editor#29]
Is it possible like how StackOverflow or the current GitHub GFM MD Editor does?

> Attach images by dragging & dropping,  selecting them, or pasting from the clipboard.

**Comment by jbt** [comment]
I was thinking about that literally this morning actually. It's something I'd very much like to add, but then it would require setting up a storage backend somewhere. That could be Amazon S3 for example, and I've already got some code written for something like that. The problem then is figuring out how to *not* make that vulnerable to abuse - I can't afford to let anyone upload anything they like to an S3 bucket under my account, because then people could use it for sharing dodgy files, or upload giant 500GB files that'd make it impossible for me to afford / maintain.

It's a tricky one. It's something I'd really like to do, but it'd need to be done right if I do it at all.

**Comment by thorsummoner** [comment]
Throw them to imgur's api c:

**Comment by jbt** [comment]
Good shout, having quickly looked at their docs it looks like it should be possible. I shall investigate :)

**Comment by praveenscience** [comment]
@jbt Can I implement it and do a pull request?

**Comment by jbt** [comment]
Yeah if you can figure it out that'd be awesome! I quite like the idea of using imgur api, at least so that it can work on https://jbt.github.io/markdown-editor without any extra backend, but maybe it can be written in such a way that the upload destination could be swapped out without changing too much.

The way I quickly had it figured in my head was something like this:

 * Some extra logic in [the `drop` evt handler](https://github.com/jbt/markdown-editor/blob/e833839d389d03eef2d1daff3354c17c7aea7620/index.html#L343-L353) that checks if the drop is a text file or an image and doing a `readAsArrayBuffer` for images. Likewise something in a `paste` event (no idea how that works with CodeMirror unfortunately).
 * Inject placeholder `![Uploading something placeholder text]()` into source text at current cursor position
 * Some kind of function like `uploadImage(fileName, arrayBuffer, contentType, callback)` where callback is `function(error, uploadedSrc){}` - then that can take care of any auth necessary for wherever the img is being uploaded, and could easily be swapped out based on storage backend
 * Once image is uploaded, go back and replace the `![Uploading something placeholder text]()` with `![fileName](uploadedSrc)`

Anyway, that's just my thoughts on how I was planning to go about implementing it, might be useful. I know JS Bin [has the same kind of functionality](https://github.com/jsbin/jsbin/blob/0922b9a5d93f9753113ef9f6ca3bd0dea70f254e/public/js/chrome/file-drop.js) - dunno if that'd be useful as inspiration

**Comment by praveenscience** [comment]
Sure... Lemme try it out. Also good idea would be `Ctrl + V` handler! What do you say?

**Comment by jbt** [comment]
Definitely - I think if you listen on the `paste` event you should be able to grab the img from there with `event.clipboardData` - [this is where jsbin does it I believe](https://github.com/jsbin/jsbin/blob/0922b9a5d93f9753113ef9f6ca3bd0dea70f254e/public/js/chrome/file-drop.js#L170)

**Comment by praveenscience** [comment]
Ah... There are a lot of tools for that! Don't worry. I'll take care... I have started working on it.

**Comment by praveenscience** [comment]
Guess I will deliver it by Sunday. Is it okay for you? I will make a pull request after that!

**Comment by soar** [comment]
Are there any news about this feature? When I have tried to Drag'n'Drop image to demo version - browser crashed.

**Comment by praveenscience** [comment]
@soar Yes, I have the problem of browser conflicts... Chrome crashes, and it doesn't work in IE.

**Comment by daimajia** [comment]
@praveenscience Any news ? :-D

**Comment by praveenscience** [comment]
@daimajia Hey buddy. I guess we all got busy with our other works. :disappointed: I will rework on this soon. :smile:

**Comment by daimajia** [comment]
@praveenscience I can understand, looking forward to the new feature. :+1:

**Comment by greenlaw110** [comment]
@praveenscience is it possible to share your existing work and I can pick up from where ever it is.

**Comment by praveenscience** [comment]
@greenlaw110 Hey that would be awesome... Will send you whatever I have. I am in a different country as when I started it before. So I don't have all the files with me.

**Comment by greenlaw110** [comment]
@praveenscience I've already done with image copy/paste, but in my project which borrow the idea from this project, but rewrite it using riot:

https://github.com/actframework/act-dockit/tree/master/src/main/resources/dockit/editor

It is part of a Java project but can be used separately if you want to write the server in whatever language.

Here are some screenshots:

![image](https://cloud.githubusercontent.com/assets/216930/15320054/d4da0d64-1c71-11e6-980f-136aa777d697.png)

![image](https://cloud.githubusercontent.com/assets/216930/15320071/eee00ab0-1c71-11e6-9665-821fcc0b4c65.png)

![image](https://cloud.githubusercontent.com/assets/216930/15320131/3ca97cd6-1c72-11e6-9973-5f95e1b49bcc.png)
